karapace/backup/api.py

```diff
@@ -338,109 +334,115 @@ def __init__(
-        topic = self.topic_name
-        with _writer(self.backup_location, overwrite=overwrite) as fp, _consumer(self.config, topic) as consumer:
+        LOG.info(
+            "Started backup in format %s of topic '%s'.",
+            version.name,
+            self.topic_name,
+        )
+
+        with _consumer(self.config, self.topic_name) as consumer:
             (partition,) = consumer.partitions_for_topic(self.topic_name)
             topic_partition = TopicPartition(self.topic_name, partition)
-            start_offset: int = consumer.beginning_offsets([topic_partition])[topic_partition]
-            end_offset: int = consumer.end_offsets([topic_partition])[topic_partition]
-            last_offset = start_offset
-            record_count = 0
-
-            fp.write(version.marker)
-            if start_offset < end_offset:  # non-empty topic
-                end_offset -= 1  # high watermark to actual end offset
-                print(
-                    (
-                        f"Started backup in format {version.name} of "
-                        f"{topic}:{partition} "
-                        f"(offset {start_offset:,} to {end_offset:,}) ..."
-                    ),
-                    file=sys.stderr,
+
+            try:
+                data_file = _write_partition(
+                    path=self.backup_location,
+                    backend=backend,
+                    consumer=consumer,
+                    topic_partition=topic_partition,
+                    poll_timeout=poll_timeout,
+                    allow_overwrite=overwrite,
                 )
-                while True:
-                    records: Collection[ConsumerRecord] = consumer.poll(poll_timeout_ms).get(topic_partition, [])
-                    if len(records) == 0:
-                        raise StaleConsumerError(topic_partition, start_offset, end_offset, last_offset, poll_timeout)
-                    for record in records:
-                        backend.store_record(fp, record)
-                        record_count += 1
-
-                    last_offset = record.offset  # pylint: disable=undefined-loop-variable
-                    if last_offset >= end_offset:
-                        break
-
-            destination = "stdout" if fp is sys.stdout else self.backup_location
-            print(
-                (
-                    f"Finished backup in format {version.name} of {topic}:{partition} "
-                    f"to {destination!r} (backed up {record_count:,} records)."
-                ),
-                file=sys.stderr,
+            except EmptyPartition:
+                LOG.warning(
+                    "Topic partition '%s' is empty, nothing to back up.",
+                    topic_partition,
+                )
+                return
+
+        end_time = datetime.datetime.now(datetime.timezone.utc)
+        if isinstance(self.backup_location, Path):
+            backend.store_metadata(
```

Comment:
This is a valid concern, but I think to properly address this we should instead write the backup to a temporary directory, and move that directory in place once done. This will change the nesting levels from current state, but IMO that makes sense.

I don't think it would be a good solution to delete the user-provided directory if we can avoid it, as it might contain other data.

Reply:
This is now addressed through the introduction of another hook on `BackupWriter`: `.prepare_location()`. Default implementation returns a no-op context manager, and for V3 we return a context manager that handles the lifecycle of a "staging directory". The staging directory behaves much like the safe writers, it moves the temporary directory in place on success, or deletes it if there are errors.